### ISSUE TITLE
Support for defining starred versions of macros

### ISSUE THREAD

**Issue by wirhabenzeit** [mathjax/MathJax#2313]
Using the `mathtools` package in LaTeX it is possible to define *paired delimiters* with automatic resizing. For example, `\DeclarePairedDelimiter{\abs}{\lvert}{\rvert}` allows to use the macros 

- `\abs{x}` = `\lvert x\rvert` 
- `\abs*{x}` = `\left\lvert x\right\rvert`
- `\abs[\Big]{x}` = `\Bigl\lvert x\Bigr\rvert`, etc.

Is it possible to replicate this behaviour in MathJax? 

So far I tried 
```
MathJax = {
   tex: {
      macros: {
             abs: ['#1\lvert #2#1\rvert', 2, '']
      }
   }
}
```
which correctly displays `\abs[\Big]{x}` and `\abs{x}`. 

Unfortunately, `\abs*{x}` is rendered as `\lvert *\rvert x` since MathJax assumes that `*` is the only argument and the optional argument is empty. Is there any possibility to separately define `\abs*{x}`?

**Comment by dpvc** [comment]
It can be done, but in order to do it, you need to define your macros using JavaScript and MathJax's extension API.  Here is an example configuration that defines the `\DeclarePairedDelimiter` control sequence so that you can define your `\abs` macro.  It also provides a configuration option where you can pre-define your paired delimiters in the configuration.

``` html
<script>
MathJax = {
  startup: {
    ready() {
      //
      //  These would be replaced by import commands if you wanted to make 
      //  a proper extension.  These are are used here for convenience
      //  just to get it running.
      //
      const Configuration = MathJax._.input.tex.Configuration.Configuration;
      const CommandMap = MathJax._.input.tex.SymbolMap.CommandMap;
      const Macro = MathJax._.input.tex.Symbol.Macro;
      const TexError = MathJax._.input.tex.TexError.default;
      const ParseUtil = MathJax._.input.tex.ParseUtil.default;
      const expandable = MathJax._.util.Options.expandable;
      
      //
      //  This is the name of the command map for the paired delimiter macros
      //
      const PAIREDMAP = 'pairedDelimitersMap';
      
      //
      //  This function implements a paried delimiter macro.
      //
      const PairedDelimiters = (parser, name, left, right) => {
        //
        //  Get the macro parameters
        //
        const star = parser.GetStar();          // true if there is a *
        const big = parser.GetBrackets(name);   // contents of optional brackets
        const math = parser.GetArgument(name);  // the math to go inside the delimiters
        //
        //  Construct the replacement string for the macro
        //
        const macro = [(star ? '\\left' : ''), (big ? big + 'l' : ''), left, '{', math, '}',
                       (star ? '\\right' : ''), (big ? big + 'r' : ''), right].join('');
        //
        //  Insert the replacement string into the TeX string, and check 
        //  that there haven't been too many maxro substitutions (prevent's 
        //  infinite loops).
        //
        parser.string = ParseUtil.addArgs(parser, macro, parser.string.slice(parser.i));
        parser.i = 0;
        if (++parser.macroCount > parser.configuration.options.maxMacros) {
          throw new TexError('MaxMacroSub1',
                             'MathJax maximum macro substitution count exceeded; ' +
                             'is there a recursive macro call?');
        }
      };

      //
      //  The is the configuration for the pariedDelimiters TeX extension.
      //
      const pariedDelimiters = Configuration.create('pairedDelimiters', {
        //
        //  Initialize the extension by creating the command map for the
        //  macros defined by \DeclarePairedDelimiter, and add the
        //  \DeclarePairedDelimiter macro itself.  Then append the
        //  command map to the given configuration as a macro handler
        //
        init(config) {
          const map = new CommandMap(PAIREDMAP, {
            DeclarePairedDelimiter: ['Declare_PairedDelimiter']
          }, {
            //
            //  Implements \DeclarePairedDelimiter control sequence.
            //
            Declare_PairedDelimiter(parser, name) {
              //
              //  Get the control sequence to define and the left- and 
              //  right-hand delimiters to use
              //
              let cs = ParseUtil.trimSpaces(parser.GetArgument(name));
              const left = parser.GetArgument(name);
              const right = parser.GetArgument(name);
              //
              //  Check that the control sequence name is valid
              //
              if (cs.charAt(0) === '\\') cs = cs.substr(1);
              if (!cs.match(/^(.|[a-z]+)$/i)) {
                throw new TexError('IllegalControlSequenceName', 'Illegal control sequence name for %1', name);
              }
              //
              //  Look up the command map and add the new macro to it using
              //  PairedDelimiters as the function and passing it the
              //  given delimiters.
              //
              const map = parser.configuration.handlers.retrieve(PAIREDMAP);
              map.add(cs, new Macro(cs, PairedDelimiters, [left, right]));
            }
          });
          config.append(Configuration.create('pairedDelimiterDefs', {handler: {macro: [PAIREDMAP]}}));
        },

        //
        //  Add any user-defined paired delimiters (from the 
        //  pairedDelimiters configuration object in the document's 
        //  option list, if any).
        //
        config(config, jax) {
          const map = jax.parseOptions.handlers.retrieve(PAIREDMAP);
          const delimiters = jax.parseOptions.options.pairedDelimiters;
          for (const cs of Object.keys(delimiters)) {
            map.add(cs, new Macro(cs, PairedDelimiters, delimiters[cs]));
          }
        },

        //
        //  Indicate that pairedDelimiters is a valid option, and can have 
        //  any number of definitions.  The format is
        //
        //      name: [left, right]
        //
        //  where 'name' is the macro name, and left and right are the 
        //  delimiters to use for the \name macro.  You can include 
        //  pre-defined macros here, which will be available without
        //  further configuration.
        //
        options: {
          pairedDelimiters: expandable({})
        }
      });
      
      MathJax.startup.defaultReady();
    }
  },
  tex: {
    packages: {'[+]': ['pairedDelimiters']},  // use the pairedDelimiters package defined above
    pairedDelimiters: {
      abs: ['\\lvert', '\\rvert']               // a user-defined paired delimiter macro
    }
  }
}
</script>
<script id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml.js"></script>
```

This configuration defines your `\abs` macro automatically, but you could also have used something like

``` tex
 \(\DeclarePairedDelimiter{\abs}{\lvert}{\rvert}\)
```

within your document to do it.

You could put this configuration into a separate file that can be loaded into multiple pages, if you wish.  Alternatively, you could build a custom extension as described in the [MathJax demos](https://github.com/mathjax/MathJax-demos-node/tree/master/custom-tex-extension) repository, and then load it like a normal MathJax extension.  You can serve it from your own site, even while taking the rest of MathJax from a CDN.

**Comment by dpvc** [comment]
The `mathtools` package is now available in v3.2, so `\DeclarePairedDelimiter` is available from there.

**Comment by mzietlow** [comment]
An important side note: the mathjax-mathtools command is `\DeclarePairedDelimiters` with trailing `s`.

I used this inside Anki-Card-Templates as follows:

```
\[
\require{mathtools}

\DeclarePairedDelimiters{\abs}{\lvert}{\rvert}
\DeclarePairedDelimiters{\norm}{\lVert}{\rVert}
\DeclarePairedDelimiters{\ang}{\langle}{\rangle}
\]
```

which worked like a charm 👍
